civiform/civiform
# When "Previous" is clicked on address correction screen, go back to block with address question

### Description

A bit of background: https://github.com/civiform/civiform/pull/6544 and https://github.com/civiform/civiform/pull/6547 updated the "Review" and "Previous" buttons on application screens to save the applicant's data or show an error modal if there are validation errors. However, the address correction view uses a different view than the typical application screen, so its "Review" and "Previous" buttons need special handling. https://github.com/civiform/civiform/pull/6589 updated the "Review" button to also save the selected suggested address when "Review" is clicked and this PR updates the "Previous" button.

This PR:
 1) Updates `AddressCorrectionBlockView` to use the new "Previous" button that saves data.
 2) In `ApplicantProgramBlocksController`, updates how the route for the "Previous" button is calculated. Because the address correction view isn't part of a block, we need to modify the block index that we use so that the user is directed to the correct previous block.

## Release notes

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### New Features


### Instructions for manual testing

#### Testing new functionality
For every test:
 - Enable the SAVE_ON_ALL_ACTIONS flag.
 - Create a program with an address question with address correction enabled.

Address matches suggestion test:
  - As an applicant, fill out the address question using the address "Address In Area, Redlands, California, 92373" so that it perfectly matches one of our fake valid addresses.
  - Click "Previous". 
  - Verify you're taken to the previous page with the address saved.

Multiple suggestions test:
  - As an applicant, fill out the address question using "Address In Area" as line 1 but an incorrect state/city/zip.
  - Click "Save&next".
  - Verify you're taken to the address correction page saying "Verify address" with a list of suggested addresses.
  - Select one of the suggestions.
  - Click "Previous".
  - Verify you're taken back to the block that had the address question. Verify that the address question is filled in with the selected suggested address saved.

No suggestions test:
 - As an applicant, fill out the address question with an address that won't trigger any suggestions (any address that doesn't use "Legit Address" or "Address In Area" as line 1 should work).
 - Click "Save&next". 
 - Verify you're taken to the address correction page saying "No valid address found".
  - Click "Previous".
  - Verify you're taken back to the block that had the address question. Verify that the address question is filled in with the original address you'd put in.

Edit test:
  - As an applicant, fill out the address correction with an address that won't trigger any suggestions.
 - Click "Save&next". 
 - On the "No valid address found" page, click Edit.
 - Verify you're taken back to the address question page.

#### Testing existing functionality

- Verify the Save&next button on the address correction page still works as before.
- Disable the SAVE_ON_ALL_ACTIONS flag and verify the "Previous" button on the address correction page still works as before (i.e., it doesn't save the chosen selection and just takes you back to the block with the address question and has the original address filled in as the answer).

### Issue(s) this completes

Part of #6450
